### ISSUE TITLE
False positive has incompatible type "Optional[str]"; expected "str"

### ISSUE THREAD

**Issue by ievgennaida** [python/mypy#11109]
**Bug Report**

Argument 2 to "getattr" has incompatible type "Optional[str]"; expected "str" mypy(error)

**To Reproduce**

Code is checking that column is not None and throws error. But it's not analyzed properly with mypy

```
if not prm or not prm.column:
    Utils.throw_server_error('Some method that always throws exception')
    # adding return here causes complains from SonarQube or pylint that can analyze 

# this line causes error because column can be None. But it cannot, because exception 
getattr(row, prm.column, None)
```

**Expected Behavior**

No warning.

**Actual Behavior**

False positive. Column property cannot be None

**Your Environment**

windows 
mypy==0.910
mypy-extensions==0.4.3
Python 3.8.10

**Comment by MatthieuDartiailh** [comment]
Did you annotate your utility function with `NoReturn` ? Trying to reproduce your issue with the following, adding `NoReturn` to stop does fix the issue.

```python
from typing import Optional, NoReturn


class A:
    pm: Optional[str]

# Removing the NoReturn leads to:
# Argument 2 to "getattr" has incompatible type "Optional[str]"; expected "str"
def stop() -> NoReturn:  
    raise RuntimeError


a = A()

if not a or not a.pm:
    stop()

getattr(a, a.pm, None)
```
